ethereum/datastructures/src/test/java/tech/pegasys/artemis/datastructures/operations/AttestationDataTest.java

```diff
@@ -61,12 +64,83 @@ void shouldNotBeProcessableBeforeFirstSlotOfTargetEpoch() {
-            new Checkpoint(UnsignedLong.ONE, Bytes32.ZERO),
+            new Checkpoint(ONE, Bytes32.ZERO),
             target);
 
     assertThat(data.getEarliestSlotForProcessing()).isEqualTo(target.getEpochStartSlot());
   }
 
+  @Test
+  public void shouldNotBeAbleToIncludeInBlockBeforeSlotAfterCreationSlot() {
+    final AttestationData data =
+        new AttestationData(
+            UnsignedLong.valueOf(60),
+            UnsignedLong.ZERO,
+            Bytes32.ZERO,
+            new Checkpoint(ONE, Bytes32.ZERO),
+            new Checkpoint(ONE, Bytes32.ZERO));
+
+    assertThat(data.canIncludeInBlockAtSlot(UnsignedLong.valueOf(60))).isFalse();
+    assertThat(data.canIncludeInBlockAtSlot(UnsignedLong.valueOf(61))).isTrue();
+  }
+
+  @Test
+  public void shouldNotBeAbleToIncludeInBlockBeforeFirstSlotOfTargetEpoch() {
+    final Checkpoint target = new Checkpoint(UnsignedLong.valueOf(10), Bytes32.ZERO);
+    final AttestationData data =
+        new AttestationData(
+            target.getEpochStartSlot().minus(ONE),
+            UnsignedLong.ZERO,
+            Bytes32.ZERO,
+            new Checkpoint(ONE, Bytes32.ZERO),
+            target);
+
+    assertThat(data.canIncludeInBlockAtSlot(target.getEpochStartSlot().minus(ONE))).isFalse();
+    assertThat(data.canIncludeInBlockAtSlot(target.getEpochStartSlot())).isTrue();
+  }
+
+  @Test
+  public void shouldNotBeAbleToIncludeInBlockTooFarAfterAttestationSlot() {
+    final AttestationData data =
+        new AttestationData(
+            UnsignedLong.valueOf(60),
+            UnsignedLong.ZERO,
+            Bytes32.ZERO,
+            new Checkpoint(ONE, Bytes32.ZERO),
+            new Checkpoint(ONE, Bytes32.ZERO));
+
+    final UnsignedLong lastValidSlot = data.getSlot().plus(UnsignedLong.valueOf(SLOTS_PER_EPOCH));
+    assertThat(data.canIncludeInBlockAtSlot(lastValidSlot)).isTrue();
+    assertThat(data.canIncludeInBlockAtSlot(lastValidSlot.plus(ONE))).isFalse();
+  }
+
+  @Test
+  public void shouldNotBeAbleToIncludeInBlockMoreThanOneEpochLater() {
+    final UnsignedLong attestationEpoch = UnsignedLong.valueOf(6);
+    final UnsignedLong nextEpoch = attestationEpoch.plus(ONE);
+    final UnsignedLong tooLateEpoch = nextEpoch.plus(ONE);
+    final AttestationData data =
+        new AttestationData(
+            compute_start_slot_at_epoch(attestationEpoch).plus(UnsignedLong.valueOf(3)),
+            UnsignedLong.ZERO,
+            Bytes32.ZERO,
+            new Checkpoint(ONE, Bytes32.ZERO),
+            new Checkpoint(ONE, Bytes32.ZERO));
+
+    // In the attestation epoch, but has to be after the attestation slot
+    assertThat(data.canIncludeInBlockAtSlot(data.getSlot().plus(ONE))).isTrue();
+
+    // Next epoch is ok
+    assertThat(data.canIncludeInBlockAtSlot(compute_start_slot_at_epoch(nextEpoch))).isTrue();
+
+    // Very last slot of next epoch is ok
+    assertThat(data.canIncludeInBlockAtSlot(compute_start_slot_at_epoch(tooLateEpoch).minus(ONE)))
+        .isFalse();
```

Comment:
this says isFalse - so it's _not_ ok? (that's different to what the comment says)

Reply:
Good spot, comment is wrong - while it's allowed in that epoch, it's more than `SLOTS_PER_EPOCH` since the attestation slot so is not ok.